### PR TITLE
Chore/vrf 585 add ctf test vrfv2plus migration

### PR DESCRIPTION
**Pull request by iljapavlovs**
1. Build chainlink image locally from latest develop
```
docker build -f ./core/chainlink.Dockerfile --build-arg=ENVIRONMENT=release --build-arg COMMIT_SHA=$(git rev-parse HEAD) --build-arg CHAINLINK_USER=chainlink -t chainlink:develop-$(git rev-parse --short HEAD)-arm .
```

2. Run the test
```
cd integration-tests 

TEST_LOG_LEVEL=debug \
SELECTED_NETWORKS=SIMULATED \
CHAINLINK_IMAGE=chainlink \
CHAINLINK_VERSION=<created image name> \
go test -v -timeout 15m -run TestVRFv2PlusMigration ./smoke
```

### CODE REVIEW

**Review by skudasov — COMMENTED** [review]
Not sure we need it for every run to pollute logs.

File: .github/workflows/integration-tests.yml [hunk 1]
```diff
@@ -241,6 +241,7 @@ jobs:
       CHAINLINK_COMMIT_SHA: ${{ github.sha }}
       CHAINLINK_ENV_USER: ${{ github.actor }}
       TEST_LOG_LEVEL: debug
+      CL_CLIENT_DEBUG: true
```
Comment (skudasov) [inline .github/workflows/integration-tests.yml:244]: Not sure if we need it enabled for all the tests, there will be a lot of logs.

Comment (iljapavlovs) [reply .github/workflows/integration-tests.yml:244]: enabled it temporarily in order to find out the failure reason as we dont have node logs available to us. will remove it

File: contracts/src/v0.8/dev/vrf/VRFConsumerBaseV2Plus.sol [hunk 2]
```diff
@@ -103,7 +103,7 @@ abstract contract VRFConsumerBaseV2Plus is IVRFMigratableConsumerV2Plus, Confirm
   error OnlyOwnerOrCoordinator(address have, address owner, address coordinator);
   error ZeroAddress();
 
-  IVRFMigratableCoordinatorV2Plus internal s_vrfCoordinator;
+  IVRFMigratableCoordinatorV2Plus public s_vrfCoordinator;
```
Comment (makramkd) [inline contracts/src/v0.8/dev/vrf/VRFConsumerBaseV2Plus.sol:106]: Does this really need to be public if subclasses are using it? It should be accessible to subclasses without changing visibility to public.

Comment (iljapavlovs) [reply contracts/src/v0.8/dev/vrf/VRFConsumerBaseV2Plus.sol:106]: this is needed in order for other coordinator to be updated during the migration process. 

this was discussed with @jinhoonbang - [slack](https://chainlink-core.slack.com/archives/C03FPM5RL9J/p1694052396948479?thread_ts=1694043917.195299&cid=C03FPM5RL9J)

Comment (jinhoonbang) [reply contracts/src/v0.8/dev/vrf/VRFConsumerBaseV2Plus.sol:106]: I think it's fine to leave it `internal` and migration will still work. 

Only reason for changing it to public was to allow querying for coordinator address. Alternatively, we could add a `getVRFCoordinator()` in base consumer or concrete consumer but I think this is fine as is

Comment (makramkd) [reply contracts/src/v0.8/dev/vrf/VRFConsumerBaseV2Plus.sol:106]: Ah got it, so this is really for external parties to query the VRF coordinator set on a consumer then. Makes sense.

**Review by kidambisrinivas — APPROVED** [review]

File: integration-tests/actions/vrfv2plus/vrfv2plus_steps.go [hunk 3]
```diff
@@ -43,90 +48,94 @@ var (
...
 func DeployVRFV2PlusContracts(
 	contractDeployer contracts.ContractDeployer,
 	chainClient blockchain.EVMClient,
+	consumerContractsAmount int,
 ) (*VRFV2PlusContracts, error) {
 	bhs, err := contractDeployer.DeployBlockhashStore()
 	if err != nil {
 		return nil, errors.Wrap(err, ErrDeployBlockHashStore)
 	}
+	//err = chainClient.WaitForEvents()
```
Comment (AnieeG) [inline integration-tests/actions/vrfv2plus/vrfv2plus_steps.go:67]: do we not need this? if no, good to remove?

File: integration-tests/actions/vrfv2plus/vrfv2plus_steps.go [hunk 4]
```diff
@@ -43,90 +48,94 @@ var (
...
+	if err != nil {
+		return nil, errors.Wrap(err, ErrWaitTXsComplete)
+	}
 	coordinator, err := contractDeployer.DeployVRFCoordinatorV2Plus(bhs.Address())
 	if err != nil {
 		return nil, errors.Wrap(err, ErrDeployCoordinator)
 	}
-	loadTestConsumer, err := contractDeployer.DeployVRFv2PlusLoadTestConsumer(coordinator.Address())
+	//err = chainClient.WaitForEvents()
+	if err != nil {
```
Comment (AnieeG) [inline integration-tests/actions/vrfv2plus/vrfv2plus_steps.go:76]: if line 75 is not needed this errcheck is duplicated

File: integration-tests/smoke/vrfv2plus_test.go [hunk 5]
```diff
@@ -108,42 +99,29 @@ func TestVRFv2PlusBilling(t *testing.T) {
 		var isNativeBilling = true
 		subNativeTokenBalanceBeforeRequest := subscription.EthBalance
 
-		jobRunsBeforeTest, err := env.CLNodes[0].API.MustReadRunsByJob(job.Job.Data.ID)
+		jobRunsBeforeTest, err := env.CLNodes[0].API.MustReadRunsByJob(vrfv2PlusData.VRFJob.Data.ID)
```
Comment (AnieeG) [inline integration-tests/smoke/vrfv2plus_test.go:102]: err check?

Comment (iljapavlovs) [reply integration-tests/smoke/vrfv2plus_test.go:102]: thanks for finding this

File: integration-tests/smoke/vrfv2plus_test.go [hunk 6]
```diff
@@ -155,4 +133,143 @@ func TestVRFv2PlusBilling(t *testing.T) {
...
+			env.GetAPIs()[0],
+			newCoordinator.Address(),
+			vrfv2PlusData.PrimaryEthAddress,
+			vrfv2PlusData.VRFKey.Data.ID,
+			vrfv2PlusData.ChainID.String(),
+			vrfv2plus_constants.MinimumConfirmations,
+		)
+		require.NoError(t, err, vrfv2plus.ErrCreateVRFV2PlusJobs)
+
+		err = vrfv2PlusContracts.Coordinator.RegisterMigratableCoordinator(newCoordinator.Address())
```
[the remark this answers is not in the document]
Comment (iljapavlovs) [reply integration-tests/smoke/vrfv2plus_test.go:201]: thanks for finding this

File: integration-tests/actions/vrfv2plus/vrfv2plus_steps.go (hunk 3 above)
Comment (iljapavlovs) [reply integration-tests/actions/vrfv2plus/vrfv2plus_steps.go]: enabled back the waitforevents

File: integration-tests/actions/vrfv2plus/vrfv2plus_steps.go (hunk 4 above)
Comment (iljapavlovs) [reply integration-tests/actions/vrfv2plus/vrfv2plus_steps.go:76]: enabled back the waitforevents

**Review by AnieeG — DISMISSED** [review]

File: integration-tests/smoke/vrfv2plus_test.go [hunk 7]
```diff
@@ -155,4 +134,144 @@ func TestVRFv2PlusBilling(t *testing.T) {
...
+		require.Equal(t, oldSubscription.EthBalance, migratedSubscription.EthBalance)
+		require.Equal(t, oldSubscription.Balance, migratedSubscription.Balance)
+		require.Equal(t, oldSubscription.Owner, migratedSubscription.Owner)
+		require.Equal(t, oldSubscription.Consumers, migratedSubscription.Consumers)
+
+		//Verify that old sub was deleted from old Coordinator
+		oldSubscription, err = vrfv2PlusContracts.Coordinator.GetSubscription(context.Background(), subID)
+		require.Error(t, err, "error not occurred when trying to get deleted subscription from old Coordinator after sub migration")
+
+		//Verify rand requests fulfills with Link Token billing
```
Comment (jinhoonbang) [inline integration-tests/smoke/vrfv2plus_test.go:254]: would be good to validate that `s_totalBalance` and `s_totalEThBalance` was incremented as expected

File: contracts/src/v0.8/dev/vrf/testhelpers/VRFCoordinatorV2PlusUpgradedVersion.sol [hunk 8]
```diff
@@ -0,0 +1,715 @@
...
+  }
+
+  function migrationVersion() public pure returns (uint8 version) {
+    return 1;
+  }
+
+  /**
+   * @inheritdoc IVRFCoordinatorV2PlusMigration
+   */
+  function onMigration(bytes calldata encodedData) external payable override {
```
Comment (jinhoonbang) [inline contracts/src/v0.8/dev/vrf/testhelpers/VRFCoordinatorV2PlusUpgradedVersion.sol:677]: I think we should also increment `s_totalBalance` and `s_totalEthBalance`

Comment (iljapavlovs) [reply contracts/src/v0.8/dev/vrf/testhelpers/VRFCoordinatorV2PlusUpgradedVersion.sol:677]: now I had a second thought on this :
should we also update following:
* `s_withdrawableTokens`, `s_withdrawableEth` with keyhash and corresponding oracle
* update `s_subscriptions` with sub id

Comment (iljapavlovs) [reply contracts/src/v0.8/dev/vrf/testhelpers/VRFCoordinatorV2PlusUpgradedVersion.sol:677]: After another thought - we dont need to update `s_withdrawableTokens`, `s_withdrawableEth` as these are updated at fulfilment phase

Comment (iljapavlovs) [reply contracts/src/v0.8/dev/vrf/testhelpers/VRFCoordinatorV2PlusUpgradedVersion.sol:677]: hmm, not sure why, but compilation fails if adding `s_subIds.add(migrationData.subId)`:

```
Error: Member "add" not found or not visible after argument-dependent lookup in struct EnumerableSet.UintSet storage ref
```

`s_subIds` has internal visibility so it should allow to do it, right?

**Review by AnieeG — APPROVED** [review]

**Review by jinhoonbang — APPROVED** [review]

**Pull request merged**